README.md
# Snowdon

Snowdon is a lightweight snowflake ID implementation in Rust. It offers a simple blocking and a lock-free thread-safe
generator implementation. Moreover, it supports custom Snowflake formats and epochs.

## Snowflake IDs

Snowflake IDs are unique IDs that include the timestamp of their "birth" (milliseconds since a fixed epoch) and a
sequence number to allow the creation of multiple snowflakes in a single millisecond. They can also include a machine
or generator ID for unique IDs across multiple generating instances without synchronization.

Twitter originally introduced snowflakes as IDs for tweets. While their implementation had a fixed epoch and snowflake
layout, most other services that adopted snowflakes have changed the epoch or layout (e.g. by using all 42 bits to store
the timestamp or by omitting the machine ID). Thus, when referring to "snowflake IDs", we mean timestamps associated
with a sequence number and, optionally, (instance-specific) constant data like a machine ID or the leading 0 in
Twitter's snowflakes.

## Project goals and non-goals

Snowdon aims to be a general-purpose snowflake ID implementation. We have the following goals:

- Provide a correct thread-safe implementation of snowflake IDs.<br/>
  The IDs generated by Snowdon should be unique, monotonic, and consistent with other machines in a deployment. You can
  learn more about correctness in the [section below](#correctness).
- Efficiently generate snowflake IDs.<br/>
  Snowflake IDs should work everywhere in an application. As this forces the ID generator to generate countless
  snowflakes, it must be as efficient as possible.
- Support custom snowflake formats.<br/>
  Unlike other ID formats, snowflakes usually differ from the original implementation, e.g. by using a different epoch
  or changing the number of bits used to store the various parts of a snowflake. Snowdon makes it easy to define and use
  these custom formats without requiring users to re-implement snowflakes from scratch.
- Be lightweight.<br/>
  Snowdon aims to be usable across various applications, so it's crucial to keep it lightweight to avoid unnecessarily
  increasing the resulting binary size or compile time.

However, we define the following non-goals to outline what Snowdon is *not*:

- Provide specific implementations for the various types of snowflakes used in existing software.<br/>
  Countless services have adopted snowflake IDs, and most of them use custom epochs, machine IDs, or even custom bit
  counts to represent the various parts of a snowflake in the 64-bit integer. Supporting specific implementations by
  providing a default implementation in Snowdon adds unnecessary clutter to the library that most users won't need.
  Instead, we aim to make custom snowflake formats as simple as possible. However, we do support the original Twitter
  layout of Snowflakes combined with a custom epoch, as many projects default to this layout.
- Provide macros or a DSL to make snowflake specifications easier to write.<br/>
  While we could create macros to remove some boilerplate from custom snowflake implementations, this would
  unnecessarily increase the compile time while only slightly reducing the overall level of boilerplate, as generators
  only need to be specified once.
- Provide coordination for multi-instance deployments.<br/>
  Snowflakes are designed to allow uncoordinated *unique* IDs by including a machine ID in each snowflake. If IDs must
  be unique across a multi-generator deployment, users must add a unique generator ID/machine ID to their snowflake
  format. It is up to the user to verify that the machine ID they pass to Snowdon generators is unique.
- Provide a means to automatically generate unique machine IDs e.g. by using the machine's IP address.<br/>
  The format of the machine ID highly depends on the application. Most simple approaches don't work for every
  deployment, and more general ways to assign machine IDs require cross-instance synchronization. Thus, it's up to the
  user to derive machine IDs in a way that suits their application.
- Provide a tool to specify snowflake layouts or epochs dynamically. Snowdon's primary focus is creating and using
  "real" snowflake IDs. Its API is designed to work with instance-wide constant layouts and epochs to avoid having to
  specify them with every call to one of the library's functions. However, this doesn't mean that all constants must
  be defined at compile time. E.g., it's fine to obtain the application's epoch at runtime before the first snowflake is
  generated.

## Usage

To use this crate, simply add the following line to your `Cargo.toml` file in the `dependencies` table:

```toml
snowdon = "^0.2"
```

If you want to use the lock-free implementation, it's recommended to disable the `blocking` feature (enabled by default)
by disabling this crate's default features:

```toml
snowdon = { version = "^0.2", default-features = false, features = ["lock-free"] }
```

To enable Serde support for snowflakes and snowflake comparators, enable the `serde` feature:

```toml
snowdon = { version = "^0.2", default-features = false, features = ["serde"] }
```

The types of this library are designed to make it hard to misuse them. Specifically, snowflakes have two type parameters
that represent their layout (the composition of a snowflake - this defines how many bits are used for the individual
parts of the snowflake) and their epoch (the timestamp that the individual timestamps in snowflakes are based on).
Because of this, it's impossible to use snowflakes in an API that expects a different layout or epoch.

Moreover, this crates doesn't provide an (easy) way to convert timestamps into snowflakes. Instead, it provides a
comparator type that can be compared with snowflakes directly. However, users of this crate can't accidentally use those
comparators in places that expect snowflakes.

Our documentation includes various examples for all complex methods. When implementing a trait, refer to the function's
documentation for detailed requirements and examples.

## SemVer compatibility

Snowdon takes Semantic Versioning compatibility seriously. In addition to the [SemVer specification][semver], we
guarantee that *patch* releases for versions starting with `0` (e.g. `0.1.1`) don't include breaking changes. I.e.,
you should be able to (automatically) update from version `0.1.0` to version `0.1.1` without having to check for
breaking changes. This behaviour effectively matches [Cargo's SemVer compatibility][cargo-semver].

[semver]: https://semver.org/

[cargo-semver]: https://doc.rust-lang.org/stable/cargo/reference/resolver.html#semver-compatibility

## Minimum Rust version policy

Our current minimum supported Rust version is `1.56.1`. We might increase our MSRV in the future, but we try to only do
so if there's a good reason for it. Only having a few dependencies that each have a stable MSRV policy, we shouldn't
need to increase our MSRV transitively because of a dependency update.

Note that benchmarks and tests require a higher Rust version, as they add dependencies that don't support the MSRV
above.

## Correctness

As detailed in [Project goals and non-goals](#project-goals-and-non-goals), Snowdon aims to be a **correct**
implementation of Twitter's Snowflake algorithm. All generators this crate implements should be safe to use across
multiple threads. More specifically, we try to ensure the following behaviour:

- All snowflakes returned by the same generator are **unique**. I.e., no snowflake will be returned twice.
- There are no "gaps" in the sequence number. If a snowflake with the sequence number `1` exists, another snowflake with
  the same timestamp and the sequence number `0` must also exist.
- Snowflakes are strictly monotonic. If the snowflake `b` was generated *after* the snowflake `a`, `a < b`.
- Snowdon uses the current system time for the snowflake. While this should be an obvious correctness requirement, all
  snowflake-generating instances in a deployment must adhere to this rule.

Note that "before" and "after" in the list above refer to the actual time. I.e., even if the system clock is not
monotonic, Snowdon only returns a snowflake `b` after producing the snowflake `a` for all subsequent invocations if
`a < b`.

In an attempt to verify the lock-free algorithm, a PROMELA implementation for it is provided in the `spin` directory.
Refer to the [README](spin/README.md) in that directory for more information about the model and our lock-free snowflake
implementation.

**Please note** that while we *try* to provide a correct and efficient implementation, there's no mathematical proof
that Snowdon's implementation is indeed correct. We do our best to ensure that our code is correct, but if you're
planning to use Snowdon in a production environment, you should verify that our implementation meets your expectations.
If you find any bugs or potential efficiency improvements, please open an issue as described
in [CONTRIBUTING.md](CONTRIBUTING.md).

## Benchmarks

Snowdon provides benchmarks in `benches`. You can run them with

```shell
cargo bench --all-features
```

When running the test on an AMD Ryzen 9 5900X with Linux 6.3.5, we get the following results:

| Benchmark                        | Average time per snowflake (in ns) | Estimated snowflake IDs per second |
|----------------------------------|------------------------------------|------------------------------------|
| Lock-free generator (sequential) | 29.188                             | 34260655                           |
| Blocking generator (sequential)  | 29.038                             | 34437633                           |
| Lock-free generator (10 threads) | 84.954                             | 11771076                           |
| Blocking generator (10 threads)  | 1594.4                             | 627195                             |

The last two entries in the table above are for an extreme scenario where ten threads concurrently generate snowflakes
in a loop. It's meant to simulate severe contention between threads, and with most applications, the actual times per
snowflake should be closer to the sequential benchmarks. Moreover, most implementations outperform the limitations of
snowflake layouts with 12 sequence number bits. Under almost all circumstances, the generators are fast enough for a
regular application's ID requirements. This includes applications that need to generate more than 100000 IDs per second.

## Licence

Copyright (C) 2023 Archer <archer@nefarious.dev>

Snowdon is licensed under either of

- Apache License 2.0 (SPDX-License-Identifier: Apache-2.0)<br/>
  [COPYING-APACHE](COPYING-APACHE) or <https://www.apache.org/licenses/LICENSE-2.0>
- MIT License (Expat License; SPDX-License-Identifier: MIT)<br/>
  [COPYING-MIT](COPYING-MIT) or <https://opensource.org/license/mit/>

at your option.

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as
defined in the Apache-2.0 licence, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Release 0.2.0

## Features

- `Cargo.toml` now correctly includes both licences (`Apache-2.0 OR MIT`) in the `license` field.
- A new `CHANGELOG.md` file lists (breaking) changes between versions.

## Breaking changes

- The default generator implementation was switched to `lock-free`.</br>
  I.e. the `blocking` feature is now turned off by default, and `lock-free` is enabled. If your code uses
  `Generator::generate_blocking()`, you'll have to enable the `blocking` feature explicitly. However, unless you want to
  keep using the blocking implementation, no changes are required if you use `Generator::generate`.
- Snowflakes are now serialized as a u64 instead of a structure.

# Release 0.1.0

The initial release of Snowdon. This release introduces the following features:

- a lock-free snowflake ID generator implementation,
- a blocking snowflake ID generator implementation,
- a snowflake comparator implementation enabling users to compare snowflakes with arbitrary timestamps,
- a flexible Snowflake type allowing static enforcement of specific layout and epoch implementations without adding any
  runtime overhead,
- and an implementation for the classic snowflake ID layout introduced by Twitter.

Moreover, this release adds unit tests, an integration test using loom, and a PROMELA implementation of our lock-free
algorithm that can be verified with SPIN.

Refer to the release's documentation for details on the introduced APIs. Happy generating!

tests/loom_generator.rs
/*
 * Copyright © 2023 Archer <archer@nefarious.dev>
 * Licensed under the Apache License, Version 2.0 (the "Licence");
 * you may not use this file except in compliance with the Licence.
 * You may obtain a copy of the Licence at
 *     https://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#[cfg(not(loom))]
compile_error!("this test requires the `loom` configuration option");

#[cfg(not(feature = "lock-free"))]
compile_error!("this test requires the `lock-free` feature");

use loom::sync::Arc;
use loom::thread;
use snowdon::system_time_mock::SystemTime;
use snowdon::{Epoch, Generator, Layout};
use std::collections::HashSet;
use std::sync::atomic::{AtomicU64, Ordering};

struct SimpleParams;

impl Layout for SimpleParams {
    fn construct_snowflake(timestamp: u64, sequence_number: u64) -> u64 {
        assert!(!Self::exceeds_timestamp(timestamp) && !Self::exceeds_sequence_number(sequence_number));
        timestamp << 22 | sequence_number
    }
    fn timestamp(input: u64) -> u64 {
        input >> 22
    }
    fn exceeds_timestamp(input: u64) -> bool {
        input >= 1 << 42
    }
    fn sequence_number(input: u64) -> u64 {
        input & ((1 << 22) - 1)
    }
    fn exceeds_sequence_number(input: u64) -> bool {
        input >= 1 << 22
    }
    fn is_valid_snowflake(_input: u64) -> bool {
        true
    }
}

impl Epoch for SimpleParams {
    fn millis_since_unix() -> u64 {
        0
    }
}

type SimpleGenerator = Generator<SimpleParams, SimpleParams>;

#[test]
fn unique_lock_free() {
    loom::model(|| {
        let _g = SystemTime::acquire_lock();
        let next_time = Arc::new(AtomicU64::new(6));
        SystemTime::set_time(5);

        let generator = Arc::new(SimpleGenerator::default());

        // Create two snowflake generating threads and a thread incrementing the current time
        let snowflakes: Vec<_> = (0..2)
            .map(|_| {
                let generator = generator.clone();
                thread::spawn(move || {
                    // Generate 3 snowflakes and return them
                    (0..3)
                        .map(|_| generator.generate_lock_free().unwrap().into_inner())
                        .collect::<Vec<_>>()
                })
            })
            .collect();
        let time = thread::spawn(move || {
            for _ in 0..2 {
                SystemTime::set_time(next_time.fetch_add(1, Ordering::Relaxed));
            }
        });
        let mut set = HashSet::with_capacity(6);
        for snowflake in snowflakes.into_iter().flat_map(|snowflake| snowflake.join().unwrap()) {
            set.insert(snowflake);
        }
        time.join().unwrap();
        assert_eq!(6, set.len());
    });
}


src/classic.rs
/*
 * Copyright © 2023 Archer <archer@nefarious.dev>
 * Licensed under the Apache License, Version 2.0 (the "Licence");
 * you may not use this file except in compliance with the Licence.
 * You may obtain a copy of the Licence at
 *     https://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

//! An implementation of the classic [`Snowflake`] layout introduced by Twitter.

use crate::{Epoch, Layout, Snowflake};
use std::marker::PhantomData;

/// A [`Layout`] implementation for the classic snowflake layout introduced by Twitter.
///
/// Snowflakes constructed with this layout consist of a leading `0` bit, 41 bits for a timestamp in milliseconds, 10
/// bits for an instance ID, and 12 bits for the sequence number. The leading `0` bit guarantees that snowflakes with
/// this layout keep their properties (namely, monotonicity) when converted into signed 64-bit integers.
///
/// Note that this layout doesn't specify the snowflake's epoch, however. Even when using this layout, you'll have to
/// specify your own epoch by implementing [`Epoch`].
///
/// # Example
///
/// ```
/// use snowdon::{
///     ClassicLayout, ClassicLayoutSnowflakeExtension, Epoch, Generator,
///     MachineId, Snowflake,
/// };
///
/// struct SnowflakeParams;
///
/// impl Epoch for SnowflakeParams {
///     fn millis_since_unix() -> u64 {
///         // The epoch used by Twitter for their snowflake IDs
///         1288834974657
///     }
/// }
///
/// impl MachineId for SnowflakeParams {
///     fn machine_id() -> u64 {
///         // Somehow obtain this machine's ID (e.g. from the private IP
///         // address or a configuration file)
/// #       0
///     }
/// }
///
/// // Make our snowflake specification available to the rest of the application
/// type MySnowflake =
///     Snowflake<ClassicLayout<SnowflakeParams>, SnowflakeParams>;
/// type MySnowflakeGenerator =
///     Generator<ClassicLayout<SnowflakeParams>, SnowflakeParams>;
///
/// // Use our snowflake format
/// let snowflake = MySnowflake::from_raw(1541815603606036480).unwrap();
/// assert_eq!(367597485448, snowflake.timestamp_raw());
/// assert_eq!(0x017A, snowflake.machine_id());
/// assert_eq!(0, snowflake.sequence_number());
/// ```
#[derive(Debug)]
pub struct ClassicLayout<I>
where
    I: MachineId,
{
    _marker: PhantomData<I>,
}

impl<I> ClassicLayout<I>
where
    I: MachineId,
{
    const TIMESTAMP_BITS: usize = 41;
    const TIMESTAMP_MASK: u64 =
        ((1 << Self::TIMESTAMP_BITS) - 1) << (Self::MACHINE_ID_BITS + Self::SEQUENCE_NUMBER_BITS);
    const MACHINE_ID_BITS: usize = 10;
    const MACHINE_ID_MASK: u64 = ((1 << Self::MACHINE_ID_BITS) - 1) << Self::SEQUENCE_NUMBER_BITS;
    const SEQUENCE_NUMBER_BITS: usize = 12;
    const SEQUENCE_NUMBER_MASK: u64 = (1 << Self::SEQUENCE_NUMBER_BITS) - 1;

    /// Returns the machine ID of the given snowflake.
    ///
    /// Usually, you shouldn't call this function directly. Instead, use
    /// [`machine_id`](ClassicLayoutSnowflakeExtension::machine_id) directly on the snowflake by importing
    /// `ClassicLayoutSnowflakeExtension`.
    #[inline]
    pub fn machine_id(input: u64) -> u64 {
        (input & Self::MACHINE_ID_MASK) >> Self::SEQUENCE_NUMBER_BITS
    }

    /// Returns whether the given machine ID exceeds the maximum supported by this layout.
    #[inline]
    fn exceeds_machine_id(input: u64) -> bool {
        input >= 1 << Self::MACHINE_ID_BITS
    }
}

impl<I> Layout for ClassicLayout<I>
where
    I: MachineId,
{
    #[inline]
    fn construct_snowflake(timestamp: u64, sequence_number: u64) -> u64 {
        let machine_id = I::machine_id();
        assert!(
            !Self::exceeds_timestamp(timestamp)
                && !Self::exceeds_sequence_number(sequence_number)
                && !Self::exceeds_machine_id(machine_id)
        );
        (timestamp << (Self::MACHINE_ID_BITS + Self::SEQUENCE_NUMBER_BITS))
            | (machine_id << Self::SEQUENCE_NUMBER_BITS)
            | sequence_number
    }

    #[inline]
    fn timestamp(input: u64) -> u64 {
        (input & Self::TIMESTAMP_MASK) >> (Self::MACHINE_ID_BITS + Self::SEQUENCE_NUMBER_BITS)
    }

    #[inline]
    fn exceeds_timestamp(input: u64) -> bool {
        input >= 1 << Self::TIMESTAMP_BITS
    }

    #[inline]
    fn sequence_number(input: u64) -> u64 {
        input & Self::SEQUENCE_NUMBER_MASK
    }

    #[inline]
    fn exceeds_sequence_number(input: u64) -> bool {
        input >= 1 << Self::SEQUENCE_NUMBER_BITS
    }

    #[inline]
    fn is_valid_snowflake(input: u64) -> bool {
        // Check whether the 64th bit is set to 0
        input < 1 << 63
    }
}

/// An extension for [`Snowflake`]s to get the snowflake's machine ID.
///
/// This trait is implemented for all snowflakes that use the [`ClassicLayout`] layout implementation.
///
/// # Example
///
/// ```
/// use snowdon::{
///     ClassicLayout, ClassicLayoutSnowflakeExtension, Generator, Snowflake,
/// };
/// # use snowdon::{Epoch, MachineId};
/// # struct SnowflakeParams;
/// # impl Epoch for SnowflakeParams {
/// #     fn millis_since_unix() -> u64 {
/// #         // The epoch used by Twitter for their snowflake IDs
/// #         1288834974657
/// #     }
/// # }
/// # impl MachineId for SnowflakeParams {
/// #     fn machine_id() -> u64 {
/// #         // Somehow obtain this machine's ID (e.g. from the private IP
/// #         // address or a configuration file)
/// #         0
/// #     }
/// # }
///
/// type MySnowflake = Snowflake<ClassicLayout<SnowflakeParams>, SnowflakeParams>;
/// type MySnowflakeGenerator =
///     Generator<ClassicLayout<SnowflakeParams>, SnowflakeParams>;
///
/// let snowflake = MySnowflake::from_raw(1541815603606036480).unwrap();
/// assert_eq!(0x017A, snowflake.machine_id());
/// assert_eq!(
///     snowflake.machine_id(),
///     ClassicLayout::<SnowflakeParams>::machine_id(snowflake.into_inner())
/// );
/// ```
pub trait ClassicLayoutSnowflakeExtension {
    /// Returns the snowflake's machine ID.
    ///
    /// Refer to the [trait documentation](Self) for an example.
    fn machine_id(&self) -> u64;

    /// Returns this snowflake as a positive signed integer.
    ///
    /// This layout guarantees that the first bit of a snowflake generated with it is `0`, so snowflakes using the
    /// `ClassicLayout` can safely be serialized as a signed 64-bit integer.
    fn into_i64(self) -> i64;
}

impl<I, E> ClassicLayoutSnowflakeExtension for Snowflake<ClassicLayout<I>, E>
where
    I: MachineId,
    E: Epoch,
{
    #[inline]
    fn machine_id(&self) -> u64 {
        ClassicLayout::<I>::machine_id(self.into_inner())
    }

    #[inline]
    fn into_i64(self) -> i64 {
        // This layout guarantees a constant `0` as the first bit. I.e., we can safely convert this to a signed integer
        // without having to worry about the resulting integer being negative.
        self.inner as i64
    }
}

/// A trait that defines a [`Snowflake`]'s constant machine ID.
///
/// This trait only requires a single associated function. I.e., there's no instance available when determining the
/// machine ID. This is an intentional design decision to discourage dynamic implementations of this value. Regardless
/// of how you determine this machine's unique ID, it's important that this function always returns the same value.
///
/// There isn't a single machine ID implementation that we could encourage here, so we're not providing an example.
/// If your implementation requires accessing remote resources to determine this machine's ID, however, you should
/// consider using the `lazy_static` crate to avoid having to re-obtain this ID for every generated snowflake.
pub trait MachineId {
    /// Returns this machine's unique ID.
    ///
    /// The returned ID must remain constant throughout the runtime of this instance.
    fn machine_id() -> u64;
}

// Skip coverage: We don't test the coverage of our unit tests
#[cfg(test)]
mod tests {
    use crate::sync::atomic::{AtomicU64, Ordering};
    use crate::sync::{Mutex, MutexGuard};
    use crate::{ClassicLayout, ClassicLayoutSnowflakeExtension, Epoch, Layout, MachineId, Snowflake};

    static MACHINE_ID: AtomicU64 = AtomicU64::new(0);
    static MACHINE_LOCK: Mutex<()> = Mutex::new(());

    struct SimpleMachineId;

    impl SimpleMachineId {
        fn set_id(id: u64) {
            MACHINE_ID.store(id, Ordering::Relaxed);
        }
        fn acquire_lock() -> MutexGuard<'static, ()> {
            MACHINE_LOCK.lock().unwrap_or_else(|e| e.into_inner())
        }
    }

    impl MachineId for SimpleMachineId {
        fn machine_id() -> u64 {
            MACHINE_ID.load(Ordering::Relaxed)
        }
    }

    // We use the type below for tests that don't need changing machine IDs to prevent having to acquire a lock on its
    // state

    struct ZeroMachineId;

    impl MachineId for ZeroMachineId {
        fn machine_id() -> u64 {
            0
        }
    }

    #[test]
    fn construct_snowflake() {
        let _g = SimpleMachineId::acquire_lock();
        SimpleMachineId::set_id(0);

        // First, verify the individual parts
        assert_eq!(
            (1 << 12) - 1,
            ClassicLayout::<SimpleMachineId>::construct_snowflake(0, (1 << 12) - 1),
            "`construct_snowflake` used an unrelated sequence number"
        );
        SimpleMachineId::set_id((1 << 10) - 1);
        assert_eq!(
            (1 << 10) - 1,
            ClassicLayout::<SimpleMachineId>::construct_snowflake(0, 0) >> 12
        );
        SimpleMachineId::set_id(0);
        assert_eq!(
            (1 << 41) - 1,
            ClassicLayout::<SimpleMachineId>::construct_snowflake((1 << 41) - 1, 0) >> 22
        );

        // Verify that the largest snowflake still has a leading 0. It's guaranteed that this is the largest snowflake,
        // as we test that the code panics for larger values with the tests below.
        SimpleMachineId::set_id((1 << 10) - 1);
        assert_eq!(
            0,
            ClassicLayout::<SimpleMachineId>::construct_snowflake((1 << 41) - 1, (1 << 12) - 1) >> 63
        );

        // Verify that this layout doesn't introduce any bits for the smallest snowflake
        SimpleMachineId::set_id(0);
        assert_eq!(0, ClassicLayout::<SimpleMachineId>::construct_snowflake(0, 0));
    }

    // Ensure that values that exceed the layout's bit counts can't be used to construct snowflakes
    #[test]
    #[should_panic]
    fn extreme_timestamp() {
        let _ = ClassicLayout::<ZeroMachineId>::construct_snowflake(1 << 41, 0);
    }

    #[test]
    #[should_panic]
    fn extreme_sequence_number() {
        let _ = ClassicLayout::<ZeroMachineId>::construct_snowflake(0, 1 << 12);
    }

    #[test]
    #[should_panic]
    fn extreme_machine_id() {
        let _g = SimpleMachineId::acquire_lock();
        SimpleMachineId::set_id(1 << 10);
        let _ = ClassicLayout::<SimpleMachineId>::construct_snowflake(0, 0);
    }

    #[test]
    fn getters() {
        assert_eq!(0, ClassicLayout::<ZeroMachineId>::timestamp((1 << 22) - 1));
        assert_eq!(
            123,
            ClassicLayout::<ZeroMachineId>::timestamp(123 << 22 | ((1 << 22) - 1))
        );
        assert_eq!((1 << 41) - 1, ClassicLayout::<ZeroMachineId>::timestamp(u64::MAX >> 1));
        assert_eq!(
            0,
            ClassicLayout::<ZeroMachineId>::sequence_number((u64::MAX << 13) >> 1)
        );
        assert_eq!(
            123,
            ClassicLayout::<ZeroMachineId>::sequence_number((u64::MAX << 13) >> 1 | 123)
        );
        assert_eq!(
            (1 << 12) - 1,
            ClassicLayout::<ZeroMachineId>::sequence_number(u64::MAX >> 1)
        );
        assert_eq!(
            0,
            ClassicLayout::<ZeroMachineId>::machine_id((u64::MAX << 23) >> 1 | ((1 << 12) - 1))
        );
        assert_eq!(
            123,
            ClassicLayout::<ZeroMachineId>::machine_id((u64::MAX << 23) >> 1 | 123 << 12 | ((1 << 12) - 1))
        );
        assert_eq!((1 << 10) - 1, ClassicLayout::<ZeroMachineId>::machine_id(u64::MAX >> 1));
    }

    #[test]
    fn snowflake_extension() {
        struct SimpleEpoch;

        impl Epoch for SimpleEpoch {
            fn millis_since_unix() -> u64 {
                0
            }
        }

        type SimpleSnowflake = Snowflake<ClassicLayout<ZeroMachineId>, SimpleEpoch>;
        assert_eq!(0, SimpleSnowflake::from_raw(0).unwrap().machine_id());
        assert_eq!(234, SimpleSnowflake::from_raw(234 << 12).unwrap().machine_id());
        assert_eq!(
            (1 << 10) - 1,
            SimpleSnowflake::from_raw(u64::MAX >> 1).unwrap().machine_id()
        );
    }
}
// End skip coverage


benches/simple_snowflake.rs
/*
 * Copyright © 2023 Archer <archer@nefarious.dev>
 * Licensed under the Apache License, Version 2.0 (the "Licence");
 * you may not use this file except in compliance with the Licence.
 * You may obtain a copy of the Licence at
 *     https://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

//! A benchmark comparing the blocking generator implementation with the lock-free implementation.
//!
//! This benchmark uses a trivial snowflake implementation to test the generator implementations without any potential
//! overhead introduced by (future versions of) the classic layout implementation.

use criterion::{black_box, criterion_group, criterion_main, Criterion};
use parking_lot::{Condvar, Mutex};
use snowdon::{Epoch, Generator, Layout, Snowflake};
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

// This gives us about 139 years until we have to change the epoch below
const TIMESTAMP_BITS: usize = 42;
// We need a lot of bits here, as the generator could easily run out of snowflakes with the classic layout on a modern
// CPU. With 22 bits, we can generate 4,194,304 snowflakes per millisecond, so we'd have to generate one snowflake every
// 4 ns to exceed this limit. When running the benchmark on my local hardware, I need ~83.5 ns for a snowflake using the
// lock-free implementation. Once processors become even faster (or atomics even cheaper), we might have to increase the
// number of bits.
const SEQUENCE_BITS: usize = 22;
const SEQUENCE_MASK: u64 = (1 << SEQUENCE_BITS) - 1;

struct MySnowflakeParams;

impl Layout for MySnowflakeParams {
    #[inline]
    fn construct_snowflake(timestamp: u64, sequence_number: u64) -> u64 {
        assert!(!Self::exceeds_timestamp(timestamp) && !Self::exceeds_sequence_number(sequence_number));
        timestamp << SEQUENCE_BITS | sequence_number
    }
    #[inline]
    fn timestamp(input: u64) -> u64 {
        input >> SEQUENCE_BITS
    }
    #[inline]
    fn exceeds_timestamp(input: u64) -> bool {
        input >= 1 << TIMESTAMP_BITS
    }
    #[inline]
    fn sequence_number(input: u64) -> u64 {
        input & SEQUENCE_MASK
    }
    #[inline]
    fn exceeds_sequence_number(input: u64) -> bool {
        input >= 1 << SEQUENCE_BITS
    }
    #[inline]
    fn is_valid_snowflake(_input: u64) -> bool {
        // We don't have any constant parts in our snowflake that we could verify
        true
    }
}

impl Epoch for MySnowflakeParams {
    #[inline]
    fn millis_since_unix() -> u64 {
        // Return the first millisecond of 2023
        1672531200000
    }
}

type MySnowflake = Snowflake<MySnowflakeParams, MySnowflakeParams>;
type MyGenerator = Generator<MySnowflakeParams, MySnowflakeParams>;

fn bench_generator(iters: u64, generator_impl: fn(&MyGenerator) -> snowdon::Result<MySnowflake>) -> Duration {
    let generator = Arc::new(MyGenerator::default());
    let start_benchmark = Arc::new((Mutex::new(false), Condvar::new()));
    // Start 20 threads that are waiting for the benchmark to start
    let threads = (0..10)
        .map(|_| {
            let generator = generator.clone();
            let start = start_benchmark.clone();
            thread::spawn(move || {
                let (start_benchmark, cvar) = &*start;
                let mut started = start_benchmark.lock();
                // Wait for the benchmark to start and immediately release the lock
                if !*started {
                    cvar.wait(&mut started);
                    drop(started);
                }
                for _ in 0..iters {
                    let _ = black_box(generator_impl(&generator).unwrap());
                }
            })
        })
        .collect::<Vec<_>>();
    let (start_benchmark, cvar) = &*start_benchmark;
    let mut start_benchmark = start_benchmark.lock();

    let start = Instant::now();
    *start_benchmark = true;
    drop(start_benchmark);
    cvar.notify_all();
    for thread in threads {
        thread.join().unwrap();
    }
    start.elapsed()
}

fn generators(c: &mut Criterion) {
    let mut group = c.benchmark_group("Generators");
    group.bench_function("blocking (10 threads)", |b| {
        b.iter_custom(|iters| bench_generator(iters, Generator::generate_blocking))
    });
    group.bench_function("lock-free (10 threads)", |b| {
        b.iter_custom(|iters| bench_generator(iters, Generator::generate_lock_free))
    });
}

fn generators_sequential(c: &mut Criterion) {
    let mut group = c.benchmark_group("Generators (sequential)");
    let generator = MyGenerator::default();
    group.bench_function("blocking", |b| b.iter(|| black_box(generator.generate_blocking())));
    // Create a fresh generator to reset the sequence number
    group.bench_function("lock-free", |b| b.iter(|| black_box(generator.generate_lock_free())));
}

criterion_group!(benches, generators, generators_sequential);
criterion_main!(benches);
